File: apps/windows/Wisp.ProjectBrowser.Contracts/NativeHighlight.cs
using System.Text.Json;

using System.Text.Json.Nodes;

namespace Wisp.ProjectBrowser.Contracts;

public sealed record NativeHighlight(string Id, string Kind, string Title, string Code, string SourceProjectId, string SourceSessionId, long CreatedAt)
{
    public bool Belongs(string project, string session) => Kind == "text" && SourceProjectId == project && SourceSessionId == session;
}
public interface INativeHighlightClient
{
    Task<NativeHighlight[]> ListAsync(string project, string session, CancellationToken token = default);
    Task<NativeHighlight> StarAsync(string project, string session, string text, CancellationToken token = default);
    Task<bool> RemoveAsync(string project, string session, string id, CancellationToken token = default);
}
public sealed class NativeHighlightClient(INativeSettingsClient transport) : INativeHighlightClient
{
    public async Task<NativeHighlight[]> ListAsync(string project, string session, CancellationToken token = default)
    {
        var result = await transport.InvokeAsync("native_conversation_panel_highlights", new JsonObject { ["session_id"] = session }, project, token).ConfigureAwait(false);
        var rows = result?.Deserialize<NativeHighlight[]>(ConversationSnapshot.JsonOptions) ?? throw new InvalidDataException("Missing highlights");
        if (rows.Any(row => !row.Belongs(project, session))) throw new InvalidDataException("Highlight scope mismatch");
        return rows;
    }
    public async Task<NativeHighlight> StarAsync(string project, string session, string text, CancellationToken token = default)
    {
        var value = await transport.InvokeAsync("native_conversation_panel_highlight_star", new JsonObject { ["session_id"] = session, ["text"] = text }, project, token).ConfigureAwait(false);
        var row = value?.Deserialize<NativeHighlight>(ConversationSnapshot.JsonOptions) ?? throw new InvalidDataException("Missing saved excerpt");
        if (!row.Belongs(project, session) || row.Code != text) throw new InvalidDataException("Saved excerpt identity mismatch");
        return row;
    }
    public async Task<bool> RemoveAsync(string project, string session, string id, CancellationToken token = default) =>
        (await transport.InvokeAsync("native_conversation_panel_highlight_remove", new JsonObject { ["session_id"] = session, ["library_item_id"] = id }, project, token).ConfigureAwait(false))?.GetValue<bool>() ?? throw new InvalidDataException("Missing removal result");
}
public static class NativeSavedExcerpt
{
    /// UTF-16 range in rendered text for WinUI highlighting; ignores whitespace like WebView.
    public static Range? Find(string text, string excerpt)
    {
        var ranges = FindAll(text, excerpt);
        return ranges.Length > 0 ? ranges[0] : null;
    }
    public static Range[] FindAll(string text, string excerpt)
    {
        var needle = new string(excerpt.Where(c => !char.IsWhiteSpace(c)).ToArray());
        if (needle.Length == 0) return [];
        var indexed = text.Select((c, index) => (c, index)).Where(pair => !char.IsWhiteSpace(pair.c)).ToArray();
        var normalized = new string(indexed.Select(pair => pair.c).ToArray());
        var result = new List<Range>();
        for (var from = 0; from <= normalized.Length - needle.Length;)
        {
            var offset = normalized.IndexOf(needle, from, StringComparison.Ordinal);
            if (offset < 0) break;
            result.Add(new Range(indexed[offset].index, indexed[offset + needle.Length - 1].index + 1));
            from = offset + needle.Length;
        }
        return result.ToArray();
    }
}

File: apps/macos/Sources/WispProjectBrowser/NativeHighlight.swift
import Foundation

/// Existing shared LibraryItem; the host returns text entries for the requested conversation only.
public struct NativeHighlight: Codable, Identifiable, Equatable, Sendable {
    public let id: String
    public let kind: String
    public let title: String
    public let code: String
    public let source_project_id: String
    public let source_session_id: String
    public let created_at: Int64
    public func belongs(project: String, session: String) -> Bool {
        kind == "text" && source_project_id == project && source_session_id == session
    }
}

public enum NativeSavedExcerpt {
    /// Character offsets in rendered text; whitespace is ignored like WebView saved marks.
    public static func range(in text: String, excerpt: String) -> Range<Int>? {
        let needle = Array(excerpt).filter { !$0.isWhitespace }
        guard !needle.isEmpty else { return nil }
        let indexed = Array(text).enumerated().filter { !$0.element.isWhitespace }
        guard indexed.count >= needle.count else { return nil }
        for start in 0...(indexed.count - needle.count) {
            if zip(indexed[start..<(start + needle.count)], needle).allSatisfy({ $0.0.element == $0.1 }) {
                return indexed[start].offset..<(indexed[start + needle.count - 1].offset + 1)
            }
        }
        return nil
    }
}
